spec.html

```diff
@@ -15478,16 +15404,72 @@ <h1>Runtime Semantics: Evaluation</h1>
+        <p>When an assignment occurs within strict mode code, it is a runtime error if _lref_ in step 1.e of the first algorithm or step 8 of the second algorithm is an unresolvable reference. If it is, a *ReferenceError* exception is thrown. The |LeftHandSideExpression| also may not be a reference to a data property with the attribute value { [[Writable]]: *false* }, to an accessor property with the attribute value { [[Set]]: *undefined* }, nor to a non-existent property of an object for which the IsExtensible predicate returns the value *false*. In these cases a *TypeError* exception is thrown.</p>
+      </emu-note>
+    </emu-clause>
+
+    <emu-clause id="sec-applynumericbinaryoperator" aoid="ApplyNumericBinaryOperator">
+      <h1>Runtime Semantics: ApplyNumericBinaryOperator ( _lval_, _opText_, _rval_ )</h1>
+      <p>The abstract operation ApplyNumericBinaryOperator takes arguments _lval_ (an ECMAScript language value), _opText_ (a sequence of Unicode code points), and _rval_ (an ECMAScript language value). It performs the following steps when called:</p>
+      <emu-alg>
+        1. If _opText_ is `+`, then
+          1. Let _lprim_ be ? ToPrimitive(_lval_).
+          1. Let _rprim_ be ? ToPrimitive(_rval_).
+          1. If Type(_lprim_) is String or Type(_rprim_) is String, then
+            1. Let _lstr_ be ? ToString(_lprim_).
+            1. Let _rstr_ be ? ToString(_rprim_).
+            1. Return the string-concatenation of _lstr_ and _rstr_.
+          1. Set _lval_ to _lprim_.
+          1. Set _rval_ to _rprim_.
+        1. NOTE: At this point, it must be a numeric operation.
+        1. Let _lnum_ be ? ToNumeric(_lval_).
+        1. Let _rnum_ be ? ToNumeric(_rval_).
+        1. If Type(_lnum_) is different from Type(_rnum_), throw a *TypeError* exception.
+        1. Let _T_ be Type(_lnum_).
+        1. If _opText_ is `**`, return ? _T_::exponentiate(_lnum_, _rnum_).
+        1. If _opText_ is `*`, return _T_::multiply(_lnum_, _rnum_).
+        1. If _opText_ is `/`, return _T_::divide(_lnum_, _rnum_).
+        1. If _opText_ is `%`, return _T_::remainder(_lnum_, _rnum_).
+        1. If _opText_ is `+`, return _T_::add(_lnum_, _rnum_).
+        1. If _opText_ is `-`, return _T_::subtract(_lnum_, _rnum_).
+        1. If _opText_ is `&lt;&lt;`, return _T_::leftShift(_lnum_, _rnum_).
+        1. If _opText_ is `&gt;&gt;`, return _T_::signedRightShift(_lnum_, _rnum_).
+        1. If _opText_ is `&gt;&gt;&gt;`, return _T_::unsignedRightShift(_lnum_, _rnum_).
+        1. If _opText_ is `&amp;`, return _T_::bitwiseAND(_lnum_, _rnum_).
+        1. If _opText_ is `|`, return _T_::bitwiseOR(_lnum_, _rnum_).
+        1. If _opText_ is `^`, return _T_::bitwiseXOR(_lnum_, _rnum_).
+        1. Assert: Execution cannot reach this step.
```

Comment:
This is a bit weird. We could just assert "opText is ^" before the previous step and remove the condition.

Reply:
We could, and have done similarly elsewhere, but the asymmetry bugs me. (If only EcmaSpeak had an exhaustive switch construct.)

Reply:
(Odd, I didn't get a notification of your comment.) Okay, I've replaced the multiple if-steps with a lookup table.